src/lowering/vtable.rs

```diff
@@ -0,0 +1,1308 @@
+//! Virtual table generation of classes and function blocks.
+//!
+//! This module is responsible for creating virtual tables for classes and function blocks, enabling
+//! dynamic dispatch for polymorphic use cases. In short, every method call needs to be invoked indirectly
+//! through a virtual table at runtime, which essentially is a collection of function pointers pointing to
+//! the method implementations of POUs. The process of creating these virtual tables can be broken down into
+//! three tasks:
+//!
+//! # 1. Virtual Table POU Member Field
+//! Every root, i.e. non-extended, class or function block will receive a `__vtable: POINTER TO __VOID` member
+//! field. For example a function block such as
+//! ```text
+//! FUNCTION_BLOCK A
+//!     VAR
+//!         one: DINT;
+//!     END_VAR
+//! END_FUNCTION_BLOCK
+//! ```
+//! will internally expand to
+//! ```text
+//! FUNCTION_BLOCK A
+//!     VAR
+//!         __vtable: POINTER TO __VOID;
+//!         one: DINT;
+//!     END_VAR
+//! END_FUNCTION_BLOCK
+//! ```
+//! Note that we use a `VOID` type here because the actual type assigned to the `__vtable` member differs from
+//! POU to POU.
+//!
+//! # 2. Virtual Table Struct Definition
+//! As mentioned, virtual tables are essentially just a collection of function pointers reflecting a POU.
+//! Consider a function block A with methods `foo` and `bar` as well as a function block B extending A with
+//! methods `foo` (inherited) and `bar` (overridden) as well as `baz` (new). For these two function blocks we
+//! would generate the following virtual table structures:
+//!
+//! ```text
+//! TYPE __vtable_A:
```

Comment:
Why not have vtable_b contain a vtable_a struct? I thought thatt was the initial plan since it can be transparent?
I'm guessing it's easier to just blindly look for the foo member?
I'm just wondering how this works with casting but i also don't think it would be an issue, we could derefernce the vtable_b as vtable_a and we get the first 2 members.

Reply:
Yes, from a desugaring POV it's much easier to just blindly call a method rather than find the correct (nested) vtable, then potentially access yet another vtable before accessing the call. Hence I opted for the flattened structure. Re casting: As long as the order is correct, casting _should_ be no issue https://github.com/PLC-lang/rusty/blob/a8533ec1523a15b1377ae2b3e97487dfbca05b13/src/lowering/polymorphism.rs#L83-L107